### PR TITLE
Vectorize multi-component mapping with numpy.select

### PR DESCRIPTION
**Pull request by tonyfujs**
## Summary

Replace row-wise `.apply(axis=1)` pattern matching in `apply_multi_component_map()` with vectorized numpy operations using `np.select()`. This improves performance by scaling with the number of rules and columns rather than the number of data rows.

## Key Changes

- **Replaced iterative row matching with vectorized boolean masks**: Build one boolean mask per rule across all rows simultaneously, then use `np.select()` to resolve them in order (first match wins).
- **Removed `re` module dependency**: Regex matching now uses pandas `.str.fullmatch()` instead of `re.fullmatch()`.
- **Added numpy import**: Required for `np.select()` and boolean array operations.
- **Improved edge case handling**: Explicitly handle the case where no rules are defined (empty `maps` list) by setting the target column to `None`.
- **Added test coverage**: Two new tests verify correct behavior when all rows are unmapped and when no rules are defined.

## Implementation Details

The new approach:
1. Iterates over rules (not rows) to build conditions
2. For each rule, constructs a boolean mask by ANDing column-wise masks
3. Uses `np.select(conditions, choices, default=None)` to apply the first matching rule to each row
4. Handles regex patterns via pandas `.str.fullmatch()` with `.fillna(False)` to convert NaN (non-string values) to False
5. Preserves rule ordering and strict column-pattern matching via `zip(..., strict=True)`

This vectorization is particularly beneficial for large datasets with many rows, as the complexity shifts from O(rows × rules × columns) to O(rules × columns × rows) with vectorized operations.

https://claude.ai/code/session_01Fc2bzBkeJk8Dcz75c5sAbr

### CODE REVIEW

File: src/tidysdmx/mapping.py [hunk 1]
```diff
@@ -207,30 +206,44 @@ def apply_multi_component_map(
...
+        mask = np.ones(n_rows, dtype=bool)
+        # strict=True preserves the contract that each rule must supply one
+        # pattern per source column.
+        for col, pattern in zip(source_cols, mv.source, strict=True):
+            col_series = result_df[col]
+            if pattern.startswith("regex:"):
+                regex = pattern.removeprefix("regex:")
+                col_mask = (
+                    col_series.astype(str)
+                    .str.fullmatch(regex)
```
[the remark this answers is not in the document]
Comment (tonyfujs) [reply src/tidysdmx/mapping.py]: Good catch on the `datetime64` nuance — confirmed that `str(pd.Timestamp("2020-01-01 00:00:00"))` yields `'2020-01-01 00:00:00'` whereas a vectorised `.astype(str)`/`.astype("string")` yields `'2020-01-01'` for midnight timestamps.

I've deliberately kept the vectorised stringification here rather than reproducing the exact per-scalar `str()` representation: the only faithful way to match scalar `str()` for every dtype is an element-wise `.map(str)`, which re-introduces the row-wise Python cost this PR is removing. SDMX source dimensions (incl. `TIME_PERIOD`) are virtually always string-typed, not `datetime64`, so the regex path operates on strings in practice.

The related missing-value concerns from the other review comments are now fixed in 106aced: the regex path uses the nullable `string` dtype with `na=False` so missing values are non-matches (instead of being coerced to `'nan'`/`'<NA>'`), and the exact path guards the comparison with `fillna(False)` so nullable dtypes no longer raise. If a `datetime64` source column with sub-day regex matching becomes a real use case, the cleanest fix would be an explicit pre-normalisation of that column to the desired string format before mapping.

---
_Generated by [Claude Code](https://claude.ai/code)_
